pkg/common-controller/snapshot_controller.go

```diff
@@ -171,53 +158,68 @@ func (ctrl *csiSnapshotCommonController) syncContent(content *crdv1.VolumeSnapsh
+		// 2) snapshot.Status.ReadyToUse is false
+		// 3) snapshot.Status.BoundVolumeSnapshotContentName is not set
+		if !utils.IsSnapshotReady(snapshot) || !utils.IsBoundVolumeSnapshotContentNameSet(snapshot) {
+			return ctrl.syncUnreadySnapshot(snapshot)
+		}
+		return ctrl.syncReadySnapshot(snapshot)
 	}
-	return ctrl.syncReadySnapshot(snapshot)
+	return nil
 }
 
-// processFinalizersAndCheckandDeleteContent processes finalizers and deletes the content when appropriate
-// It checks if contents exists, it checks if snapshot has bi-directional binding, it checks if
-// finalizers should be added or removed, and it checks if content should be deleted and deletes it
-// if needed.
-func (ctrl *csiSnapshotCommonController) processFinalizersAndCheckandDeleteContent(snapshot *crdv1.VolumeSnapshot) error {
-	klog.V(5).Infof("processFinalizersAndCheckandDeleteContent VolumeSnapshot[%s]: %s", utils.SnapshotKey(snapshot), utils.GetSnapshotStatusForLogging(snapshot))
-
+func (ctrl *csiSnapshotCommonController) checkContentAndBoundStatus(snapshot *crdv1.VolumeSnapshot) (*crdv1.VolumeSnapshotContent, bool, bool, error) {
 	// If content is deleted already, remove SnapshotBound finalizer
-	content, err := ctrl.contentExists(snapshot)
-	if err != nil {
-		return err
+	content, err := ctrl.getContentFromStore(snapshot)
+	if err != nil || content == nil {
+		return nil, false, false, err
 	}
 	deleteContent := false
-	// It is possible for contentExists to return nil, nil
-	if content != nil && content.Spec.DeletionPolicy == crdv1.VolumeSnapshotContentDelete {
+	// It is possible for getContentFromStore to return nil, nil
+	if content.Spec.DeletionPolicy == crdv1.VolumeSnapshotContentDelete {
 		klog.V(5).Infof("processFinalizersAndCheckandDeleteContent: Content [%s] deletion policy [%s] is delete.", content.Name, content.Spec.DeletionPolicy)
 		deleteContent = true
 	}
 
 	snapshotBound := false
 	// Check if the snapshot content is bound to the snapshot
-	if content != nil && utils.IsSnapshotBound(snapshot, content) {
+	if utils.IsSnapshotBound(snapshot, content) {
 		klog.Infof("syncSnapshot: VolumeSnapshot %s is bound to volumeSnapshotContent [%s]", snapshot.Name, content.Name)
 		snapshotBound = true
+
 	}
+	return content, deleteContent, snapshotBound, nil
+}
 
-	klog.V(5).Infof("processFinalizersAndCheckandDeleteContent[%s]: delete snapshot content and remove finalizer from snapshot if needed", utils.SnapshotKey(snapshot))
-	err = ctrl.checkandRemoveSnapshotFinalizersAndCheckandDeleteContent(snapshot, content, deleteContent)
+// processIfDeletionTimestampSet processes finalizers and deletes the content when appropriate
+// It checks if contents exists, it checks if snapshot has bi-directional binding, it checks if
+// finalizers should be removed, and it checks if content should be deleted and deletes it if needed.
+func (ctrl *csiSnapshotCommonController) processIfDeletionTimestampSet(snapshot *crdv1.VolumeSnapshot) error {
```

Comment:
I think the logic implemented here is confusing. There are several layers of calls. Is the following logic sufficient?

if snapshot.DeletionTimeStamp is set (it indicates it must has finalizer)
1. if snapshotAsSource finalizer is set
 --  (double check pvc is still using it) -- do nothing
 -- if pvc finishes restoring, remove finalizer
2. if SnapshotBound finalizer is set
  -- content exist, set deletiontimestamp on content (can also set annotation at the same time here)
  -- context does not exist, delete finalizer

Reply:
There are 3 steps in processIfDeletionTimestampSet (renamed to processSnapshotWithDeletionTimestamp):
1. Call a helper function checkContentAndBoundStatus() to check content and bound status.
2. Call checkandRemoveSnapshotFinalizersAndCheckandDeleteContent() with information obtained from step 1.  This function name is very long but the name suggests what it does.  It determines whether to remove finalizers on snapshot and whether to delete content.
3. Call checkandRemovePVCFinalizer() to determine whether to remove the finalizer on PVC.

I added comments to explain that.